package/libpiksi/libpiksi/src/loop.c

```diff
@@ -0,0 +1,467 @@
+    piksi_log(LOG_ERR, "Failed to allocate callback context for signal handler add");
+    goto failure;
+  }
+
+  return uv_signal;
+
+failure:
+  pk_loop_destroy_uv_handle((uv_handle_t *)uv_signal);
+  return NULL;
+}
+
+/**
+ * @brief uv_loop_poll_handler - wrapping callback for uv_poll_t
+ * @param poller: poll handle
+ * @param status: poll operation status
+ * @param events: event that triggered this callback
+ */
+static void uv_loop_poll_handler(uv_poll_t *poller, int status, int events)
+{
+  uv_handle_t *handle = (uv_handle_t *)poller;
+  pk_loop_t *loop = pk_loop_from_uv_handle(handle);
+  pk_callback_ctx_t *cb_ctx = pk_callback_context_from_uv_handle(handle);
+
+  if (status < 0) {
+    piksi_log(LOG_ERR, "UV_ERROR %s", uv_strerror(status));
+    return;
+  }
+  if (events & UV_DISCONNECT) {
+    piksi_log(LOG_ERR, "uv_poll_event - UV_DISCONNECT");
+    return;
+  }
+
+  if (cb_ctx->callback != NULL) {
+    cb_ctx->callback(loop, handle, cb_ctx->data);
+  }
+}
+
+void * pk_loop_endpoint_reader_add(pk_loop_t *pk_loop,
+                                   pk_endpoint_t *pk_ept,
+                                   pk_loop_cb callback,
+                                   void *context)
+{
+  assert(pk_loop != NULL);
+  assert(pk_ept != NULL);
+  assert(callback != NULL);
+
+  uv_poll_t *uv_poll = (uv_poll_t *)malloc(sizeof(uv_poll_t));
+  if (uv_poll == NULL) {
+    piksi_log(LOG_ERR, "Failed to allocate uv_poll for reader add");
+    goto failure;
+  }
+
+  int poll_handle = pk_endpoint_poll_handle_get(pk_ept);
+  if (uv_poll_init_socket(pk_loop->uv_loop, uv_poll, poll_handle) != 0) {
+    piksi_log(LOG_ERR, "Failed to init uv_poll for reader add");
+    goto failure;
+  }
+
+  if (uv_poll_start(uv_poll, UV_READABLE | UV_DISCONNECT, uv_loop_poll_handler) != 0) {
+    piksi_log(LOG_ERR, "Failed to start uv_poll for reader add");
+    goto failure;
+  }
+
+  if (pk_loop_add_handle_context((uv_handle_t *)uv_poll, callback, context) != 0) {
+    piksi_log(LOG_ERR, "Failed to allocate callback context for reader add");
+    goto failure;
+  }
+
+  return uv_poll;
+
+failure:
+  pk_loop_destroy_uv_handle((uv_handle_t *)uv_poll);
+  return NULL;
+}
+
+/**
+ * @brief match_handle_walk_callback - maps over handles attempting to match
+ * This is mostly a helper for handle_is_valid to receive the match context and
+ * use the contents to either match the current handle and call the callback or
+ * pass
+ * @param handle:
+ * @param arg:
+ */
+static void match_handle_walk_callback(uv_handle_t *handle, void *arg)
+{
+  match_handle_ctx_t *ctx = (match_handle_ctx_t *)arg;
+  if (ctx->handle == (void *)handle) {
+    if (ctx->callback != NULL) {
+      ctx->callback(handle, ctx->data);
+    }
+    ctx->match = true;
+  }
+}
+
+bool pk_loop_handle_is_valid(pk_loop_t *pk_loop, void *handle)
+{
+  assert(pk_loop != NULL);
+  match_handle_ctx_t ctx = {
+    .handle = handle,
+    .match = false,
+    .callback = NULL,
```

Comment:
Originally I was going to use this to match and remove handles to try and avoid the case where an invalid handle was passed to `pk_loop_remove_handle` but didn't end up going through with it. I can just remove the entire structure, since `pk_loop_handle_is_valid` is unused as well.